river/neighbors/knn_classifier.py

```diff
@@ -1,176 +1,193 @@
+        softmax: bool = False,
     ):
         super().__init__(
             n_neighbors=n_neighbors,
             window_size=window_size,
-            leaf_size=leaf_size,
-            p=p,
-            **kwargs
+            min_distance_keep=min_distance_keep,
+            distance_func=distance_func,
         )
         self.weighted = weighted
-        self.classes_: typing.Set = set()
-        self.kwargs = kwargs
+        self.cleanup_every = cleanup_every
+        self.classes = set()
+        self.softmax = softmax
+        self._cleanup_counter = cleanup_every
 
-    def _unit_test_skips(self):
-        return {"check_emerging_features", "check_disappearing_features"}
+    @property
+    def _multiclass(self):
+        return True
 
-    def learn_one(self, x, y):
-        """Update the model with a set of features `x` and a label `y`.
+    def clean_up_classes(self) -> "KNNClassifier":
+        """
+        Classes that are added (and removed) from the window may no longer be valid.
+        This method cleans up the window and and ensures only known classes
+        are added, and we do not consider "None" a class. It is called every
+        `cleanup_every` step, or can be called manually.
+        """
+        self.classes = {x for x in self.window if x[0][1] is not None}
+
+    def learn_one(self, x: dict, y=None):
+        """Learn a set of features `x` and optional class `y`.
+        Parameters:
+            x: A dictionary of features.
+            y: A class (optional if known).
+        Returns:
+            self
+        """
+        # Only add the class y to known classes if we actually add the point!
+        if self.nn.update((x, y), n_neighbors=self.n_neighbors):
+            self.classes.add(y)
 
-        Parameters
-        ----------
-        x
-            A dictionary of features.
-        y
-            The class label.
+        # Ensure classes known to instance reflect window
+        self._run_class_cleanup()
+        return self
 
-        Returns
-        -------
-            self
+    def _run_class_cleanup(self):
+        """
+        Helper function to run class cleanup, accounting for _cleanup_counter.
+        """
+        # clean up classes every cleanup_every steps
+        if self.cleanup_every:
+            self._cleanup_counter -= 1
+            if self._cleanup_counter == 0:
+                self.clean_up_classes()
+                self._cleanup_counter = self.cleanup_every
 
-        Notes
-        -----
-        For the K-Nearest Neighbors Classifier, fitting the model is the
-        equivalent of inserting the newer samples in the observed window,
-        and if the size_limit is reached, removing older results.
+        return self
 
+    def predict_one(self, x: dict):
+        """Predict the label of a set of features `x`.
+        Parameters:
+            x: A dictionary of features.
+        Returns:
+            The neighbors
         """
+        return self.predict_proba_one(x)
```

Comment:
This is wrong. `predict_one` returns a class, not a dict of probabilities. This method is anyway inherited by `base.Classifier`, so you don't have to implement it.